compiler/lib-wasm/generate.ml

```diff
@@ -204,13 +210,14 @@ module Generate (Target : Target_sig.S) = struct
               if b
               then return (W.Call (f, List.rev (closure :: acc)))
               else
-                match funct with
-                | W.RefFunc g ->
+                match funct, kind with
+                | W.RefFunc g, _ ->
                     (* Functions with constant closures ignore their
                        environment. In case of partial application, we
                        still need the closure. *)
-                    let* cl = if exact then Value.unit else return closure in
+                    let* cl = if exact_call kind then Value.unit else return closure in
                     return (W.Call (g, List.rev (cl :: acc)))
+                | _, Known g -> return (W.Call (g, List.rev (closure :: acc)))
```

Comment:
can we have `f <> g` here ?

Reply:
Sure. For instance, if you have a call `StringMap.iter h m`, `f` will be bound to the expression which accesses the relevant field of module `StringMap` while `g` will be bound to the actual closure (assuming that the functor application has not been optimized away).